src/RestClient.Net.DependencyInjection/RestClientExtensions.cs

```diff
@@ -7,23 +7,38 @@ namespace RestClient.Net
 {
     public static class RestClientExtensions
     {
+        private static IServiceCollection AddRestClient_Base(IServiceCollection serviceCollection)
+        {
+            serviceCollection
+                .AddSingleton<CreateHttpClient>((sp) =>
+                {
+                    var microsoftHttpClientFactoryWrapper = new MicrosoftHttpClientFactoryWrapper(sp.GetRequiredService<IHttpClientFactory>());
+                    return microsoftHttpClientFactoryWrapper.CreateClient;
+                })
+                .AddSingleton<CreateClient>((sp) =>
+                {
+                    var clientFactory = new ClientFactory(
+                        sp.GetRequiredService<CreateHttpClient>(),
+                        sp.GetService<ILoggerFactory>());
+
+                    return clientFactory.CreateClient;
+                });
+
+            return serviceCollection;
+        } 
+        
         public static IServiceCollection AddRestClient(this IServiceCollection serviceCollection, Action<CreateClientOptions>? configureClient = null)
         {
-            _ = serviceCollection
-            .AddSingleton<CreateHttpClient>((sp) =>
-            {
-                var microsoftHttpClientFactoryWrapper = new MicrosoftHttpClientFactoryWrapper(sp.GetRequiredService<IHttpClientFactory>());
-                return microsoftHttpClientFactoryWrapper.CreateClient;
-            })
-            .AddSingleton<CreateClient>((sp) =>
-            {
-                var clientFactory = new ClientFactory(
-                    sp.GetRequiredService<CreateHttpClient>(),
-                    sp.GetService<ILoggerFactory>());
+            AddRestClient_Base(serviceCollection);
+            serviceCollection.AddSingleton(sp => sp.GetRequiredService<CreateClient>()("RestClient", configureClient));
 
-                return clientFactory.CreateClient;
-            })
-            .AddSingleton((sp) => sp.GetRequiredService<CreateClient>()("RestClient", configureClient));
+            return serviceCollection;
+        }        
+        
+        public static IServiceCollection AddRestClient(this IServiceCollection serviceCollection, Action<CreateClientOptions, IServicesProvider>? configureClient = null)
+        {
+            AddRestClient_Base(serviceCollection);
+            serviceCollection.AddSingleton(sp => sp.GetRequiredService<CreateClient>()("RestClient", options => configureClient(options, sp)));
```

Comment:
@DrLeh I understand the issue you are trying to address here but this code simply won't work. 

The delegate `CreateClient` only takes one parameter of type `CreateClientOptions`. It does not take `IServiceProvider`

Please take some time to fix this up, get it to build, and test it.

Thanks

Reply:
`CreateClient` is not accepting the service provider, the provided `Action` is. Rather than passing the action as a parameter to the `CreateClient`, this is creating a new delegate, where the provided `Action` is invoked within, so this does indeed compile as we are still sending the same parameter of `Action<options>` to the `CreateClient`

I pushed an additional commit that refactors this a bit to reduce ambiguity with the nullable parameters. The code compiles on my machine and the dependency-injection related unit tests.

However, I have not yet found a way that I could register multiple rest clients under different names and then resolve them, the way you can do with `IHttpClientFactory`. I created overloads to register the clients by name but we would need some method to resolve them by name.